src/ml/neural_net/combine_base.hpp

```diff
@@ -0,0 +1,253 @@
+
+template <typename T, typename U>
+class MapPublisher;
+
+template <typename T, typename U>
+class Transform;
+
+/**
+ * Simple type expressing how many values a Subscriber is ready to receive
+ * from its Publisher.
+ */
+class Demand {
+ public:
+  static Demand Unlimited() { return Demand(-1); }
+  static Demand None() { return Demand(0); }
+
+  /** Any negative value is interpreted as "unlimited". */
+  explicit Demand(int max) : max_(max) {}
+
+  bool IsUnlimited() const { return max_ < 0; }
+  bool IsNone() const { return max_ == 0; }
+
+  /** Returns a negative number to indicate "unlimited." */
+  int max() const { return max_; }
+
+  /** Additively combines another Demand value into this one. */
+  Demand& Add(Demand other) {
+    if (IsUnlimited() || other.IsUnlimited()) {
+      max_ = -1;
+    } else {
+      max_ += other.max_;
+    }
+    return *this;
+  }
+
+  /** Decrease this demand by one if the current max is positive and finite. */
+  Demand& Decrement() {
+    if (max_ > 0) {
+      --max_;
+    }
+    return *this;
+  }
+
+ private:
+  int max_ = 0;
+};
+
+/**
+ * Interface for objects that Publishers send to Subscribers to allow the
+ * Subscribers to (potentially asynchronously) control the flow of values that
+ * the Subscriber receives from the Publisher.
+ */
+class Subscription {
+ public:
+  virtual ~Subscription() = default;
+
+  /**
+   * Requests the Publisher to stop sending anything to the Subscriber.
+   *
+   * After receiving Cancel() from a Subscriber, a Publisher should thereafter
+   * ignore all future messages from that Subscriber, including future calls to
+   * Cancel.
+   *
+   * Publishers must support Subscribers calling Cancel() from inside
+   * Subscriber::Receive(...).
+   */
+  virtual void Cancel() = 0;
+
+  /**
+   * Requests the Publisher to send the indicated number of values to the
+   * Subscriber.
+   *
+   * Publishers must support Subscribers calling Request(Demand) from inside
+   * Subscriber::Receive(Subscription), but Subscribers should avoid calling
+   * Request(Demand) inside Subscriber::Receive(Input). Instead, they should
+   * send additional Demand via the return value of Subscriber::Receive(Input)
+   * (to help prevent infinite recursion).
+   */
+  virtual void Request(Demand demand) = 0;
+};
+
+/**
+ * Type representing a message from a Publisher to a Subscriber indicating that
+ * the Subscriber will no longer receive any further messages.
+ */
+class Completion {
+ public:
+  /** Returns an instance that signals successful completion. */
+  static Completion Finished() { return Completion(); }
+
+  /**
+   * Returns an instance that signals failure, described by the given
+   * exception.
+   */
+  static Completion Failure(std::exception_ptr e) { return Completion(e); }
+
+  bool IsFinished() const { return failure_ == nullptr; }
+
+  /** Returns the exception if a failure and a null pointer otherwise. */
+  std::exception_ptr failure() const { return failure_; }
```

Comment:
it's better to use boost::optional

Reply:
This portion of the codebase is not taking a dependency on boost for now. Might adopt std::optional if we go to C++17, or write a simple implementation ourselves.

Reply:
but our codebase already used boost 1.68. Exception is better than segfault.

Reply:
What segfault do you mean? We are using `std::exception_ptr` explicitly to capture exceptions and support those in client code

Reply:
I mean one can use std::exception_ptr with null value.

Reply:
Yes, in this case, a null exception pointer corresponds to the no-error case, when a stream finished normally